Hooks/ilhooks.md

```diff
@@ -1,23 +1,369 @@
+they are just a few rules of thumb. Just make sure to keep the stack balanced.
+
+Let's see how to use `EmitDelegate` in action. Suppose we want to modify the drop platform time based on the player's
+current health. Here is one way to accomplish that using `EmitDelegate`.
+
+```csharp
+if (cursor.TryGotoNext
+(
+    i => i.MatchLdcR4(0.7f),
+    i => i.MatchNewobj<WaitForSeconds>()
+))
+{
+    // move between the ldc.r4 and the newobj; we want to have the original time on the stack as an input
+    cursor.GotoNext();
+    cursor.EmitDelegate<Func<float, float>>(f => {
+        return f + (0.5f * PlayerData.instance.GetInt(nameof(PlayerData.health)));
+    });
+}
+```
+
+
+## Debugging Broken IL
+
+IL hooks can be incredibly difficult to debug. This is because in an IL hook, if you make an error, you'll usually
+end up with invalid code, and the compiler has no way to warn you since you've made all the modifications at runtime.
+This manifests itself as an `InvalidProgramException`, which often provides very little context on what the error is.
+In many cases, the best chance you have to fix broken IL is to get out a pen and paper and work your way through the
+state of the stack as instructions run, until you get to the code you've modified. However, there are a few common
+errors that are typically caused when IL hooking, and being able to identify them can speed up your debugging
+significantly.
+
+The most common error to see, as we've discussed several times in this documentation already, is an unbalanced stack.
+This can be caused in several ways, most commonly pushing additional data onto the stack without consuming it
+appropriately or adding a delegate with arguments that doesn't push back onto the stack. If there are too many values 
+on the stack, you'll usually see an error message like "Invalid IL code in (the method you hooked): (some IL address): 
+ret" or "Invalid IL code in (the method you hooked): (some IL address): (some instruction close after where you 
+modified code)," depending on if you added or removed values from the stack respectively.
+
+Another common cause of errors is putting invalid data type on the stack (for instance, loaded a float when you should
+have loaded an int). This will look quite similar to the error above, where some code near where you made your 
+modification will show as invalid. This is because that code is expecting a certain type of input still; in the case 
+above you are providing no input, in this case you are providing the incorrect input.
+
+A final, less common cause of errors is creating invalid branch instructions. This can happen when modifying code
+after the end of an if block, or when trying to emit if/branch instructions in IL. Typically, the error is caused
+when you branch into the middle of a series of operations, so the stack is not in an expected state. This will look
```

Comment:
I made a remark on this on the following line - tldr removing a branch target is not usually the actual problem, ILCursor seems like it will remake the branch to point at the instruction at the same index. Typically, the actual problem when removing a branch target is that you have an unbalanced stack

Reply:
I wanted it cuz this happened once https://discord.com/channels/879125729936298015/879130756146954240/933744868558274581

Reply:
Weird. I specifically tried to reproduce this error and could not. I still think the "in some cases removing a branch target can cause problems" clause will suffice but let me investigate more.

Reply:
Reproduced, clarified messaging further.